### ISSUE TITLE
WEBGL_Multiview default frameBuffer

### ISSUE THREAD

**Issue by shahmeeresmail** [KhronosGroup/WebGL#2167]
Could we modify the spec to allow for this extension to be used with any texture type as a target instead of just a texture array backed default frame buffer or offscreen texture array? Specifically, we should allow for the default frame buffer to exist as either a textureArray or a texture with left and right eyes rendered side by side. There are a few use cases and reasons for this:
1. For stereo contexts, the browser has the option of using any implementation it chooses for the default frame buffer (side-by-side texture or textureArray)
2. If the extension has been implemented in a driver or in ANGLE but the browser does not support stereo contexts or it is still in development, WebGL developers can implement their projects using the default frame buffer as it exists today in order to immediately take advantage of these optimizations. There is no need to wait for the browser to support stereo buffers. When full stereo contexts are supported in the browser, using them would simply involve adding the appropriate context creation parameter but the rest of the code should function as is.
3. Implementors of the standard should have the flexibility to use as many implementations as possible under the hood and declaring that the default frame buffer for a stereo context must exist as a texture array puts restrictions on what implementations can be used.

Lastly, great job @Oletus with creating the proposed spec for this extension. It will be great to have something like this for developers to use.

**Comment by Oletus** [comment]
I think there might be some misunderstanding at play here. The intention of the wording of the proposal is that the default stereo framebuffer can be implemented in various ways under the hood - either as a texture with the left and right viewports side-by-side, as a texture array, or as a native stereo framebuffer. Is there something that prevents this in the current proposal?

I don't think that making the extension dependent on stereo contexts is a problem. Basic support for stereo contexts is a much simpler feature to implement than the extension itself. Side-by-side rendering into a mono framebuffer is a short-term hack, and won't be optimal for some devices where it will require an extra copy to get side-by-side framebuffer data onto the display. Because of this we should send a clear message that stereo context is the recommended way going forward, and not muddle it by having an alternative special side-by-side path.

**Comment by shahmeeresmail** [comment]
Thanks for clarifying that the extension would allow for any implementation of a stereo buffer.

Putting a dependency on stereo buffers can be a problem if issues are encountered during implementation or integration into browsers.  You say it is simple but it is impossible to predict what kinds of issues will be encountered during development and we should not prevent WebGL developers from using the extension until that support is there.

Side-by-side rendering is NOT a hack. In fact, some stereoscopic rendering pipelines have vertical side-by-side framebuffers which the display controller scans from directly and as a result, do not have this copy you speak of. 

Stereo contexts are the recommended way forward and by enabling the extension to be used without them we aren't sending the message it is ok but simply giving them an option until support is there for all browsers.  Making the switch to stereo contexts, when they are available will only involve adding the 'stereo' parameter to the list of context parameters.  Enabling developers to hit the ground running is very valuable.

Further, by enabling developers to use the extension before stereo contexts are available, we give the community an opportunity to test the extension using real world scenarios and to help us identify and resolve bugs early in the process.

**Comment by msc-** [comment]
> On 2016 Nov 23, at 10:27, Shahmeer Esmail <notifications@github.com> wrote:
> 
> Side-by-side rendering is NOT a hack. In fact, some stereoscopic rendering pipelines have vertical side-by-side framebuffers which the display controller scans from directly and as a result, do not have this copy you speak of.

Isn’t this then a stereo context? And “vertical side-by-side” is an oxymoron. Do you mean vertical or side by side?

Some other points:

I am struggling to see the relevance of the draw buffer additions to this extension. The underlying native extension does not use the default framebuffer and as it is using a texture array target draw buffer seems irrelevant.

I fervently hope we never see a WebGL stereo context. It must be multiview. 8 views are more is not uncommon for “auto-stereoscopic” displays.

This extension adds to the long list of reasons why backing store/canvas creation should be divorced from WebGL context creation.

Regards

    -Mark

**Comment by Oletus** [comment]
The draw buffer additions to the extension are there so that a zero-copy display pipeline can be implemented across VR display devices, which might require the framebuffer to be in various different formats. Some kind of an opaque stereo framebuffer abstraction is required for this purpose, and after a bit of discussion it seemed most fitting to extend the default framebuffer into that.

I can see the appeal of divorcing backing store/canvas creation from WebGL context creation, but as far as I'm aware the other known use cases for that have been addressed by offscreen canvas. As such I think we'd be taking on quite a lot of extra work and delays for limited payoff if we made divorcing backing store creation from context creation a part of this project.

The design of WebVR hard-codes stereo instead of allowing for an arbitrary number of viewpoints. Because of this I think that stereo support is also a reasonable addition to WebGL, instead of trying to add support for more generic multiview right from the start. A more generic mechanism for a multiview framebuffer could exist alongside stereo if such is needed in the future. As of yet the need isn't really there.

While it's true that sometimes unforeseen difficulties appear in the middle of development, I don't think we should worry about that too much beforehand. If stereo canvas will indeed prove to be harder to implement than it seems, we can consider amending the spec at that point. I also don't think that switching from non-stereo (side-by-side) to stereo canvas can be made quite as simple for applications as just switching the "stereo" flag - it will at least require changing the canvas dimensions as well.

**Comment by shahmeeresmail** [comment]
@msc- What I meant by vertical side-by-side is a single rendertarget with the left eye using the top half and the right eye using the bottom half. The side-by-side part conveys spatial locality and the vertical part conveys the direction :)

**Comment by shahmeeresmail** [comment]
There is common ground here and in fact we agree on certain things:
- The default frame buffer should be opaque as far as how it is implemented. It could be implemented as stereo texture arrays, ordinary textures, double wide textures or in some other fashion. Regardless of how it is implemented, interactions with it shouldn't change.
- Developers need an interface to tell WebGL which eye(s) to render to and not have to worry about the underlying implementation of the default frame buffer.
- Developers should be "guided" to take the optimal path for rendering multiple views whenever possible and we need to avoid allowing them to take a suboptimal whenever possible.

One way to do this is to require multiview to be used with a stereo canvas UNLESS stereo canvases are not supported in which case it can operate but with some caveats.

In other words: 
- if there is support for a stereo canvas in the browser, multiview should ONLY operate if the context was created with the stereo flag
- If there is no support for a stereo canvas in the browser, multiview should operate but perhaps with some caveats which could be documented in the spec as well. For example, the spec could state that multiview would implicitly divide the default frame buffer into left and right halves which would be the targets for all multivew calls.  Further, it could be required that the canvas be created with the stereo parameter (even though it is mostly ignored) just to ensure that once stereo buffers are implemented, they will be used by default.

I believe this would give us the best of both worlds.

**Comment by msc-** [comment]
> On 2016 Nov 24, at 1:15, Olli Etuaho <notifications@github.com> wrote:
> 
> The draw buffer additions to the extension are there so that a zero-copy display pipeline can be implemented across VR display devices, which might require the framebuffer to be in various different formats. Some kind of an opaque stereo framebuffer abstraction is required for this purpose, and after a bit of discussion it seemed most fitting to extend the default framebuffer into that.
> 
> I can see the appeal of divorcing backing store/canvas creation from WebGL context creation, but as far as I'm aware the other known use cases for that have been addressed by offscreen canvas. As such I think we'd be taking on quite a lot of extra work and delays for limited payoff if we made divorcing backing store creation from context creation a part of this project.
> 

I don’t think the use case of rendering to multiple canvases on a page from a single context has been addressed. However I agree separation would be taking on a lot of extra work. I am more expressing regret that this much cleaner way of handling all these cases was not chosen. I find the offscreen canvas mechanism particularly convoluted and dislike that the only thing preventing copying of canvases is competition between browser vendors.

As for drawBuffer, the spec says

"When both the left and right buffers are chosen for rendering, framebuffer writes will be done as if the framebuffer had two-layer texture arrays attached by framebufferTextureMultiviewWEBGL. The left buffer corresponds to gl_ViewID_OVR zero, and the right buffer corresponds to gl_ViewID_OVR one. If drawBuffer is called with the buf parameter set to either BACK_LEFT or BACK_RIGHT, only one of the stereo buffers will again be chosen for rendering.”

When only one buffer is selected by drawBuffer, how does that relate to writes using gl_ViewID_OVR? E.g., if BACK_LEFT is selected are writes to gl_ViewID_OVR 0 no-ops? This needs to be specified.

I don’t like the mixing of two mechanisms for selecting the buffers to write to. The alternative is a special gl.renderbufferStorage call that would provide “storage" which is the actual multiview hardware buffers so you can have a zero-copy pipeline. But that is difficult to reconcile with having the context and canvas created together.

Regards

    -Mark

**Comment by jdashg** [comment]
We're expecting to expose multiview through OVR_multiview2.
